### PR TITLE
make ImmutableLinkedHashMap serializable

### LINKED ISSUE #944: java.io.NotSerializableException on  play.api.libs.json.ImmutableLinkedHashMap
We have a redis cache implementation on our scala project to perform caching of some objects.
It has an api similar to PlayCache api and it worked with play-json 2.9.4. 
Internally we use the java serialization (via ObjectInputStream/ObjetOutputStream)  during call to redis set/get operations.
We are migrating to latest Play 2.9.0 version and the play-json 2.10.2 . And after migration during serialization we got following exception if serialized object contains JsObject as it's member  (or child member contains JsObject...)
Can you mark play.api.libs.json.ImmutableLinkedHashMap as serializable ?

```
Caused by: java.io.NotSerializableException: play.api.libs.json.ImmutableLinkedHashMap
at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1187)
at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1572)
at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1529)
at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1572)
at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1529)
at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1572)
at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1529)
at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
at java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:350)
at io.waylay.redis.RedisCache$.anySerialize(RedisCache.scala:43)
at io.waylay.redis.RedisCache$.io$waylay$redis$RedisCache$$$anonfun$vSerializer$1(RedisCache.scala:26)
at io.waylay.redis.RedisCache$$anonfun$2.serialize(RedisCache.scala:26)
at redis.api.strings.Set.<init>(Strings.scala:125)
at redis.commands.Strings.set(Strings.scala:81)
at redis.commands.Strings.set$(Strings.scala:76)
at redis.RedisClient.set(Redis.scala:78)
at io.waylay.redis.RedisCache.set(RedisCache.scala:64)
```

### PR DESCRIPTION
**Pull request by ramazanyich**
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [ ] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Fixes

Fixes #944 

## Purpose

In order to be able to serialize objects that contain JsObject as a member need to have internal underlying ImmutableLinkedHashMap implement serializable interface

### CODE REVIEW

**Comment by ramazanyich** [comment]
have you an idea how I can test the serialization in scala-native.
Looks like problem is in fact that ObectInputStream/ObjectOutputStream is not included into javalibs of scala-native.
which gives a problem during test/NativeLink

**Comment by jtjeferreira** [comment]
> have you an idea how I can test the serialization in scala-native. Looks like problem is in fact that ObectInputStream/ObjectOutputStream is not included into javalibs of scala-native. which gives a problem during test/NativeLink

I think it is better to just have that test as JVM test in `play-json/jvm/src/test/scala/play/api/libs/json`

File: play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala
Comment (jtjeferreira) [inline play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala:23]: If you look at std lib collections you will notice the use of `scala.collection.generic.DefaultSerializable`

File: play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala
Comment (jtjeferreira) [inline play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala:25]: again, by looking at std lib collections you can see the use of `@SerialVersionUID` annotation instead of the field...

**Comment by ramazanyich** [comment]
> > have you an idea how I can test the serialization in scala-native. Looks like problem is in fact that ObectInputStream/ObjectOutputStream is not included into javalibs of scala-native. which gives a problem during test/NativeLink
> 
> I think it is better to just have that test as JVM test in `play-json/jvm/src/test/scala/play/api/libs/json`

thanks for the tip

File: play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala
Comment (ramazanyich) [reply play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala:26]: adjusted

File: play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala
Comment (ramazanyich) [reply play-json/shared/src/main/scala-2.13-/play/api/libs/json/ImmutableLinkedHashMap.scala:24]: DefaultSerializable available only with Scala2.13 and serialization tests failing with ClassCastException then..

File: play-json/shared/src/main/scala/play/api/libs/json/Reads.scala
Comment (jtjeferreira) [inline play-json/shared/src/main/scala/play/api/libs/json/Reads.scala:35]: This is a sign there is some reference to a Reads that is leaking, and is not a good sign... I checked the branch locally and the problem is in JsNull. If you add a `@transient` to the `@transient implicit val reads` and remove this code the test is green in scala 2.12...

Comment (ramazanyich) [reply play-json/shared/src/main/scala/play/api/libs/json/Reads.scala]: done

**Review by jtjeferreira — APPROVED** [review]
LGTM, but would be better if a maintainer could have a look

**Comment by ramazanyich** [comment]
@mkurz can you take a look on this PR

**Review by mkurz — APPROVED** [review]
Should work

**Comment by mkurz** [comment]
@Mergifyio backport main

**Pull request merged**
